### DISCUSSION TITLE
Feature request: Shell script as type of action

### DISCUSSION THREAD

**Discussion by judoole** [community/community#26031] [Actions]
<p>In short, the ability to write actions using shell script as well as todays node and docker.</p>
<p>Why? The containers that runs the actions has a lot installed that could very well be used in small shell scripted actions. That way the actions would be fast as they would need no download time.</p>
<p>Also I think this would be easier, or a good alternative, to work with than the node type. I find the commiting of the node_modules a bit cumbersome and makes the repos bloated.</p>

**Comment by BelHFL** [comment]
<p>Hi <a class="mention" href="/judoole">@judoole</a>,</p>
<p>Thanks for this feedback! We’re always working to improve GitHub and the GitHub Community Forum, and we consider every suggestion we receive. I’ve logged your feature request in our internal feature request list. Though I can’t guarantee anything or share a timeline for this, I can tell you that it’s been shared with the appropriate teams for consideration.</p>
<p>Cheers!</p>

**Comment by eine** [comment]
<blockquote>
<p><a class="mention" href="/judoole">@judoole</a> wrote:<br>
I find the commiting of the node_modules a bit cumbersome and makes the repos bloated.</p>
</blockquote>
<p>You can use a packager for this. E.g.: <a href="https://github.com/1138-4EB/issue-runner/blob/master/webpack.config.js" rel="nofollow noopener">https://github.com/1138-4EB/issue-runner/blob/master/webpack.config.js</a></p>

**Comment by bcomnes** [comment]
<p>Likewise, lots of hoops to jump through just to run some simple cleanup scripts.</p>

**Comment by bcomnes** [comment]
<p>I wrote my first action which boils down to shelling out from node to a bash script.  Lots of hoops to jump through, and even more if I wanted to use the core modules (setting up a build step)!  </p>
<p><a href="https://github.com/bcomnes/cleanup-xvfb" rel="nofollow noopener">https://github.com/bcomnes/cleanup-xvfb</a> if anyone is curious.  Overall though, I’m loving actions.  It seems as if shelling out from node seems to work, a script entrypoint could easily be exposed in the <code>using</code> filed:</p>
<pre><code class="lang-auto">
runs:

&amp;nbsp; using: sh

&amp;nbsp; main: some-script.sh

</code></pre>

**Comment by peter-evans** [comment]
<p>Very much agree with this. I would also like to see other action runners like Python, for example. There doesn’t even need to be a toolkit for these other languages. Someone would probably create one using the javascript actions/toolkit as a reference implementation.</p>
<p>I would just like to be able to say this in action.yml:</p>
<pre><code>runs:
  using: 'python'
  main: 'main.py'
</code></pre>
<p>Also, one of the issues right now is that container actions only work on linux VMs. The only option for a multi-platform action is javascript.</p>

**Comment by DannyBen** [comment]
<p>I was also wondering the same thing.</p>
<p>The ability to build simple shell based actions would be great, and the ability to build actions in Ruby would be awesome.</p>

**Comment by DannyBen** [comment]
<p>If anyone is interested, I have created an easy workaround until GitHub Actions provide support for simple shell actions.</p>
<p>The idea is this:</p>
<ol>
<li>
<p>Have a small “run-remote” script in the repo (for example, in the .github folder)</p>
</li>
<li>
<p>This script accepts two arguments: GitHub repo and path to script in that repo</p>
</li>
<li>
<p>You can then simply run it in any step, and execute remote, github-hosted shell scripts</p>
</li>
</ol>
<p>This is the .github/remote-run script (should have the executable bit on):</p>
<pre><code>#!/usr/bin/env bash
repo=$1
script=$2
shift 2
url="https://raw.githubusercontent.com/$repo/master/$script"
curl -s $url | bash -s $*
</code></pre>
<p>Here is a sample step:</p>
<pre><code>steps:
- name: Checkout code
  uses: actions/checkout@v2

- name: Run some remote script
  run: .github/remote-run dannyben/shell-action-example say-hello Thor
</code></pre>
<p>Here is the proof of concept and more details:</p>
<p><a href="https://github.com/DannyBen/shell-action-example" rel="nofollow noopener">https://github.com/DannyBen/shell-action-example</a></p>

**Comment by judoole** [comment, marked as answer]
<p>Guess this is now solved with <a href="https://docs.github.com/en/actions/creating-actions/creating-a-composite-run-steps-action" rel="nofollow noopener">https://docs.github.com/en/actions/creating-actions/creating-a-composite-run-steps-action</a></p>
